src/analysis/processing/qgsalgorithmzonalhistogram.cpp

```diff
@@ -0,0 +1,323 @@
+    int maxRow = ( int )( ( mExtent.yMaximum() - intersectRect.yMinimum() ) / mRasterUnitsPerPixelY ) + 1;
+
+    nCellsX = maxColumn - offsetX;
+    nCellsY = maxRow - offsetY;
+
+    //avoid access to cells outside of the raster (may occur because of rounding)
+    if ( ( offsetX + nCellsX ) > mNbCellsXProvider )
+    {
+      nCellsX = mNbCellsXProvider - offsetX;
+    }
+    if ( ( offsetY + nCellsY ) > mNbCellsYProvider )
+    {
+      nCellsY = mNbCellsYProvider - offsetY;
+    }
+
+    QHash< double, qgssize > fUniqueValues;
+    middlePoints( featureGeometry, offsetX, offsetY, nCellsX, nCellsY, fUniqueValues );
+
+    if ( fUniqueValues.count() < 1 )
+    {
+      //the cell resolution is probably larger than the polygon area. We switch to slower precise pixel - polygon intersection in this case
+      preciseIntersection( featureGeometry, offsetX, offsetY, nCellsX, nCellsY, fUniqueValues );
+    }
+
+    for ( auto it = fUniqueValues.constBegin(); it != fUniqueValues.constEnd(); ++it )
+    {
+      if ( uniqueValues.indexOf( it.key() ) == -1 )
+      {
+        uniqueValues << it.key();
+      }
+      featuresUniqueValues[ zonesFieldIndex > -1 ? f.attribute( zonesFieldIndex ) : QVariant( f.id() )][it.key()] += it.value();
+    }
+
+    current++;
+  }
+
+  std::sort( uniqueValues.begin(), uniqueValues.end() );
+
+  QString fieldPrefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+  QgsFields newFields;
+  for ( auto it = uniqueValues.constBegin(); it != uniqueValues.constEnd(); ++it )
+  {
+    newFields.append( QgsField( QStringLiteral( "%1%2" ).arg( fieldPrefix ).arg( *it ), QVariant::LongLong, QString(), -1, 0 ) );
+  }
+  QgsFields fields = QgsProcessingUtils::combineFields( zones->fields(), newFields );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields,
+                                          zones->wkbType(), zones->sourceCrs() ) );
+
+  it = zones->getFeatures( QgsFeatureRequest() );
+  while ( it.nextFeature( f ) )
+  {
+    QgsAttributes attributes = f.attributes();
+    QHash< double, qgssize > fUniqueValues = featuresUniqueValues.value( zonesFieldIndex > -1 ? f.attribute( zonesFieldIndex ) : QVariant( f.id() ) );
+    for ( auto it = uniqueValues.constBegin(); it != uniqueValues.constEnd(); ++it )
+    {
+      attributes += fUniqueValues.value( *it, 0 );
+    }
+
+    QgsFeature outputFeature;
+    outputFeature.setGeometry( f.geometry() );
+    outputFeature.setAttributes( attributes );
+
+    sink->addFeature( outputFeature, QgsFeatureSink::FastInsert );
+  }
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+  return outputs;
+}
+
+void QgsZonalHistogramAlgorithm::middlePoints( const QgsGeometry &poly, int pixelOffsetX, int pixelOffsetY, int nCellsX, int nCellsY, QHash< double, qgssize > &uniqueValues )
```

Comment:
is this the same as the zonal stats middlePoints method? If so, would be good to reuse that to avoid the duplicate code.

Reply:
@nyalldawson , it's actually not. 

That said, I'll probably clean up the zone stats code a bit following this PR.